libs/utils/include/celix_properties.h

```diff
@@ -17,86 +17,474 @@
+                                                         const char* file,
+                                                         const char* header);
 
-CELIX_UTILS_EXPORT const char* celix_properties_get(const celix_properties_t *properties, const char *key, const char *defaultValue);
+/**
+ * @brief Get the entry for a given key in a property set.
+ *
+ * @param[in] properties The property set to search.
+ * @param[in] key The key to search for.
+ * @return The entry for the given key, or a NULL if the key is not found.
+ */
+CELIX_UTILS_EXPORT celix_properties_entry_t* celix_properties_getEntry(const celix_properties_t* properties,
+                                                                       const char* key);
 
-CELIX_UTILS_EXPORT void celix_properties_set(celix_properties_t *properties, const char *key, const char *value);
+/**
+ * @brief Get the value of a property.
+ *
+ * @param[in] properties The property set to search.
+ * @param[in] key The key of the property to get.
+ * @param[in] defaultValue The value to return if the property is not set.
+ * @return The value of the property, or the default value if the property is not set.
+ */
+CELIX_UTILS_EXPORT const char*
+celix_properties_get(const celix_properties_t* properties, const char* key, const char* defaultValue);
 
-CELIX_UTILS_EXPORT void celix_properties_setWithoutCopy(celix_properties_t *properties, char *key, char *value);
+/**
+ * @brief Get the type of a property value.
+ *
+ * @param[in] properties The property set to search.
+ * @param[in] key The key of the property to get the type of.
+ * @return The type of the property value, or CELIX_PROPERTIES_VALUE_TYPE_UNSET if the property is not set.
+ */
+CELIX_UTILS_EXPORT celix_properties_value_type_e celix_properties_getType(const celix_properties_t* properties,
+                                                                          const char* key);
 
-CELIX_UTILS_EXPORT void celix_properties_unset(celix_properties_t *properties, const char *key);
+/**
+ * @brief Set the value of a property.
+ *
+ * @param[in] properties The property set to modify.
+ * @param[in] key The key of the property to set.
+ * @param[in] value The value to set the property to.
+ * @return CELIX_SUCCESS if the operation was successful, CELIX_ENOMEM if there was not enough memory to set the entry.
+ */
+CELIX_UTILS_EXPORT celix_status_t celix_properties_set(celix_properties_t* properties,
+                                                       const char* key,
+                                                       const char* value);
 
-CELIX_UTILS_EXPORT celix_properties_t* celix_properties_copy(const celix_properties_t *properties);
+/**
+ * @brief Set the value of a property without copying the value string.
+ *
+ * @param[in] properties The property set to modify.
+ * @param[in] key The key of the property to set. This string will be used directly, so it must not be freed or
```

Comment:
`key ` is actually moved into properties. In case of error, it is freed. Thus it should not be touched after this function returns.

Reply:
We also need to emphasize that `key` will be freed by `free`. It can not come from arbitrary  memory allocator.

Reply:
updated the documentation to clarify this